File: src/assets/admin/cgi/editorjs.py
#!/usr/bin/python3

'''
Created: 2023.12.06
Author: Vladimir Vons <VladVons@gmail.com>
License: GNU, see LICENSE for more details
'''


import os
import cgi
#
from lib import Run, ParseQuery, GetPostAsData, GetPostAsJson, LoadFileJson, SaveFileJson, HttpAnswer, SaveFileStorage


cFileDat = 'editorjs.dat.json'

def Main():
    Code = 200
    Res = {'status': 'ok'}

    Query = os.environ.get('QUERY_STRING')
    Query = ParseQuery(Query)
    aMode = Query.get('mode')
    if (aMode == 'save_editor'):
        try:
            Data = GetPostAsJson()
            SaveFileJson(cFileDat, Data.get('data'))
        except Exception as E:
            Code = 500
            Res = {'status': 'err', 'info': str(E)}
    elif (aMode == 'load_editor'):
        GetPostAsData()
        Res = LoadFileJson(cFileDat)
    elif (aMode == 'save_img'):
        Form = cgi.FieldStorage()
        if ('image' in Form):
            DirRoot = os.environ.get('DOCUMENT_ROOT')
            Dir = os.environ.get('SCRIPT_NAME').rsplit('/', maxsplit=1)[0]
            try:
                File = SaveFileStorage(Form['image'], f'{DirRoot}{Dir}')
                Res = {
                    'success' : 1,
                    'file': {'url' : f'{Dir}/{File}'}
                }
            except Exception as E:
                Res = {'success' : 0, 'error': str(E)}
    else:
        Code = 404
        Res = {'status': 'err', 'info': f'unknown mode `{aMode}`'}

    HttpAnswer(Res, Code)

Run(Main)

File: src/assets/admin/js/editorjs.js
/*
Created: 2023.12.06
Author: Vladimir Vons <VladVons@gmail.com>
License: GNU, see LICENSE for more details
*/


function editorJsInit() {
    const urlRoot = 'assets/admin/cgi/editorjs.py'

    const editor = new EditorJS({
        holder: 'editor_js',
        autofocus: true,
        tools: {
            header: {
                class: Header,
                inlineToolbar: ['link']
            },
            image: {
                class: ImageTool,
                inlineToolbar: true,
                config: {
                    endpoints: {
                        byFile: urlRoot + '?mode=save_img'
                    }
                },
                tunes: ['alignment']
            },
            list: {
                class: List,
                inlineToolbar: ['link'],
                tunes: ['alignment']
            },
            linkTool: {
                class: LinkTool,
                config: {
                    endpoint: 'http://localhost:8008/fetchUrl',
                }
            },
            code: {
                class: CodeTool
            },
            table: {
                class: Table
            },
            paragraph: {
                class: Paragraph,
                inlineToolbar: true,
                tunes: ['alignment']
            },
            alignment: {
                class: AlignmentBlockTune,
                config:{
                    default: "left",
                    blocks: {
                        header: 'center',
                        list: 'right'
                    }
                }
            }
        }
    })

    const elEditorBtnSave = document.getElementById('viEditorBtnSave')
    elEditorBtnSave.addEventListener('click', function(aEvent) {
        editor.save().then((aEditorData) => {
            const url = urlRoot + '?mode=save_editor'
            postJson(url, {'data': aEditorData})
            .then(data => {
                if (data.status == 'err') {
                    showTooltip('error ' + data.info)
                }else{
                    showTooltip(`saved with ${url}. status ${data.status}`)
                }
            })
        })
    })

    const elEditorBtnLoad = document.getElementById('viEditorBtnLoad')
    elEditorBtnLoad.addEventListener('click', function(aEvent) {
        const url = urlRoot + '?mode=load_editor'
        postJson(url, {})
        .then(data => {
            if (Object.keys(data).length) {
                editor.isReady.then(() => {
                    editor.render(data)
                    showTooltip('loaded with ' + url)
                })
            }else{
                showTooltip('no saved data')
            }
        })
    })
}

editorJsInit()
